core/src/main/java/feign/Contract.java

```diff
@@ -211,99 +207,238 @@ protected void nameParam(MethodMetadata data, String name, int i) {
     }
   }
 
-  class Default extends BaseContract {
+  /**
+   * {@link Contract} base implementation that works by declaring witch annotations should be
+   * processed and how each annotation modifies {@link MethodMetadata}
+   */
+  public abstract class DeclarativeContract extends BaseContract {
+
+    Map<Class<Annotation>, ClassAnnotationProcessor<Annotation>> classAnnotationProcessors =
+        new HashMap<>();
+    private Map<Class<Annotation>, MethodAnnotationProcessor<Annotation>> methodAnnotationProcessors =
+        new HashMap<>();
+    Map<Class<Annotation>, ParameterAnnotationProcessor<Annotation>> parameterAnnotationProcessors =
+        new HashMap<>();
 
-    static final Pattern REQUEST_LINE_PATTERN = Pattern.compile("^([A-Z]+)[ ]*(.*)$");
+    public final List<MethodMetadata> parseAndValidatateMetadata(Class<?> targetType) {
+      // any implementations must register processors
+      return super.parseAndValidatateMetadata(targetType);
+    }
 
+    /**
+     * Called by parseAndValidateMetadata twice, first on the declaring class, then on the target
+     * type (unless they are the same).
+     *
+     * @param data metadata collected so far relating to the current java method.
+     * @param clz the class to process
+     */
     @Override
-    protected void processAnnotationOnClass(MethodMetadata data, Class<?> targetType) {
-      if (targetType.isAnnotationPresent(Headers.class)) {
-        String[] headersOnType = targetType.getAnnotation(Headers.class).value();
-        checkState(headersOnType.length > 0, "Headers annotation was empty on type %s.",
-            targetType.getName());
-        Map<String, Collection<String>> headers = toMap(headersOnType);
-        headers.putAll(data.template().headers());
-        data.template().headers(null); // to clear
-        data.template().headers(headers);
-      }
+    protected final void processAnnotationOnClass(MethodMetadata data, Class<?> targetType) {
+      Arrays.stream(targetType.getAnnotations())
+          .forEach(annotation -> classAnnotationProcessors
+              .getOrDefault(annotation.annotationType(), ClassAnnotationProcessor.DO_NOTHING)
+              .process(annotation, data));
```

Comment:
This feels odd to me.  I can see that you are going for functional style, but the empty function feels forced.  This can be accomplished with a simple loop and lookup and that doesn't require the empty function:

```java
for (Annotation annotation : targetType.getAnnotations()) {
   AnnotationProcessor processor = this.classAnnotationProcessors.get(annotation);
   if (processor != null) {
      process.process(annotation, data);
   }
}
```

Can you help me understand if this is more than just a style choice?

Reply:
100% pure stylish choice.

Original change had this if to check for null.

Not having it allows me to do a leaner lambda function.

Reply:
I don't normally like to haggle over style, but I really thing that if we don't have a processor, we shouldn't create an empty one just to avoid a `null` check.  If you really want to do this, I suggest the following change:

```java
private Optional<AnnotationProcessor> getProcessor(Annotation annotation) {
   return Optional.ofNullable(this.processors.get(annotation));
}
```
then you can do this:
```java
Arrays.stream(targetType.getAnnotations())
   .forEach(annotation -> {
      getProcessor(annotation)
         .ifPresent(processor -> processor.process(annotation, data))
  });
```
keeps the functional style without the empty lambda.  Thoughts?

Reply:
Take a look on the new approach using a guarded function.

Reply:
An alternative approach that doesn't require the empty lambda is to use `Optional` and use `ifPresent`.

```java
private Optional<AnnotationProcessor> getAnnotationProcessor(
   Annotation annotation) {
      return Optional.ofNullable(this.processors.get(annotation));
}
```
```java
Arrays.stream(target.getAnnotations())
   .forEach(annotation -> {
      getAnnotationProcessor(annotation)
         .ifPresent(processor -> processor.process(annotation, data));
   });
```

This removes the need for the empty lambda.  What do you think?